From 7355bb75eb20ced35416798ca331fa5e6ffab18d Mon Sep 17 00:00:00 2001
From: Jens Petersen <petersen@redhat.com>
Date: Tue, 28 Jun 2022 05:50:03 +0800
Subject: [PATCH] 'temporary' ephemeral distrobox (#335) (#336)

* initial poc of 'temporary' ephemeral distrobox (#335)

installing basic packages a little time but
this is already quite useful to me.

This could surely be improved more...
In the longer term the right thing may well be to fork
this off to a separate file like the other cmds.

* ephemeral: create dedicated script
* ephemeral: tidy up help, add usage docs

Signed-off-by: Luca Di Maio <luca.dimaio1@gmail.com>

Co-authored-by: Luca Di Maio <luca.dimaio1@gmail.com>
(cherry picked from commit 8cbe2a491b606f4bc074c1c0e330d5cc2524ab55)
Signed-off-by: Nathan Chancellor <nathan@kernel.org>
---
 distrobox                         |   6 +-
 distrobox-ephemeral               | 134 ++++++++++++++++++++++++++++++
 docs/README.md                    |   1 +
 docs/usage/distrobox-ephemeral.md |  41 +++++++++
 docs/usage/usage.md               |   1 +
 5 files changed, 182 insertions(+), 1 deletion(-)
 create mode 100755 distrobox-ephemeral
 create mode 100644 docs/usage/distrobox-ephemeral.md

diff --git a/distrobox b/distrobox
index 5cbb6de..665edfe 100755
--- a/distrobox
+++ b/distrobox
@@ -38,9 +38,10 @@ distrobox version: ${version}
 Choose one of the available commands:
 	create
 	enter
+	ephemeral
 	list
-	stop
 	rm
+	stop
 	version
 EOF
 }
@@ -72,6 +73,9 @@ case "${distrobox_command}" in
 	rm)
 		"${distrobox_path}"/distrobox-rm "$@"
 		;;
+	ephemeral*)
+		"${distrobox_path}"/distrobox-ephemeral "$@"
+		;;
 	-V | --version | version)
 		printf "distrobox: %s\n" "${version}"
 		exit 0
diff --git a/distrobox-ephemeral b/distrobox-ephemeral
new file mode 100755
index 0000000..53a519f
--- /dev/null
+++ b/distrobox-ephemeral
@@ -0,0 +1,134 @@
+#!/bin/sh
+# SPDX-License-Identifier: GPL-3.0-only
+#
+# This file is part of the distrobox project:
+#    https://github.com/89luca89/distrobox
+#
+# Copyright (C) 2021 distrobox contributors
+#
+# distrobox is free software; you can redistribute it and/or modify it
+# under the terms of the GNU General Public License version 3
+# as published by the Free Software Foundation.
+#
+# distrobox is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with distrobox; if not, see <http://www.gnu.org/licenses/>.
+
+# POSIX
+# Optional env variables:
+#	DBX_CONTAINER_MANAGER
+#   DBX_CONTAINER_NAME
+#   DBX_NON_INTERACTIVE
+
+# Dont' run this command as sudo.
+if [ "$(id -u)" -eq 0 ]; then
+	printf >&2 "Running %s as sudo is not supported.\n" "$(basename "${0}")"
+	printf >&2 " try instead running:\n"
+	printf >&2 "	%s --root %s\n" "$(basename "${0}")" "$*"
+	exit 1
+fi
+
+container_command=""
+create_flags=""
+distrobox_path="$(dirname "${0}")"
+extra_flags=""
+rootful=0
+verbose=0
+version="1.3.2"
+
+# Print usage to stdout.
+# Arguments:
+#   None
+# Outputs:
+#   print usage with examples.
+show_help() {
+	cat << EOF
+distrobox version: ${version}
+
+Usage:
+
+	distrobox-ephemeral --name container-name [--force]
+
+Options:
+
+	--root/-r:		launch podman/docker with root privileges. Note that if you need root this is the preferred
+				way over "sudo distrobox"
+	--verbose/-v:		show more verbosity
+	--help/-h:		show this message
+	--/-e:			end arguments execute the rest as command to execute at login	default: bash -l
+	--version/-V:		show version
+
+See also:
+
+	distrobox-create --help
+
+	for a list of additional flags to specify during the creation.
+	distrobox-ephemeral can use all flags from distrobox-create.
+
+EOF
+}
+
+# Parse arguments
+while :; do
+	case $1 in
+		-h | --help)
+			# Call a "show_help" function to display a synopsis, then exit.
+			show_help
+			exit 0
+			;;
+		-r | --root)
+			shift
+			rootful=1
+			;;
+		-v | --verbose)
+			verbose=1
+			shift
+			;;
+		-V | --version)
+			printf "distrobox: %s\n" "${version}"
+			exit 0
+			;;
+		-e | --exec | --)
+			shift
+			container_command="-- $*"
+			break
+			;;
+		*) # Default case: If no more options then break out of the loop.
+			# If we have a flagless option and container_name is not specified
+			# then let's accept argument as container_name
+			if [ -n "$1" ]; then
+				create_flags="${create_flags} $1"
+				shift
+			else
+				break
+			fi
+			;;
+	esac
+done
+
+set -o errexit
+set -o nounset
+# set verbosity
+if [ "${verbose}" -ne 0 ]; then
+	set -o xtrace
+	extra_flags="${extra_flags} --verbose"
+fi
+
+# prepend sudo if we want podman or docker to be rootful
+if [ "${rootful}" -ne 0 ]; then
+	extra_flags="${extra_flags} --root"
+fi
+
+name=$(  mktemp -u distrobox-XXXXXXXXXX)
+# shellcheck disable=SC2086
+"${distrobox_path}"/distrobox-create --name ${name} --yes ${create_flags} ${extra_flags}
+# shellcheck disable=SC2086
+"${distrobox_path}"/distrobox-enter "${name}" ${container_command} ${extra_flags}
+# shellcheck disable=SC2086
+"${distrobox_path}"/distrobox-stop "${name}" --yes ${extra_flags}
+# shellcheck disable=SC2086
+"${distrobox_path}"/distrobox-rm --force "${name}" --yes ${extra_flags}
diff --git a/docs/README.md b/docs/README.md
index df00785..80983cc 100644
--- a/docs/README.md
+++ b/docs/README.md
@@ -47,6 +47,7 @@ graphical apps (X11/Wayland), and audio.
   - [Outside the distrobox](usage/usage.md#outside-the-distrobox)
     - [distrobox-create](usage/distrobox-create.md)
     - [distrobox-enter](usage/distrobox-enter.md)
+    - [distrobox-ephemeral](usage/distrobox-ephemeral.md)
     - [distrobox-list](usage/distrobox-list.md)
     - [distrobox-rm](usage/distrobox-rm.md)
     - [distrobox-stop](usage/distrobox-stop.md)
diff --git a/docs/usage/distrobox-ephemeral.md b/docs/usage/distrobox-ephemeral.md
new file mode 100644
index 0000000..21cee7b
--- /dev/null
+++ b/docs/usage/distrobox-ephemeral.md
@@ -0,0 +1,41 @@
+<!-- markdownlint-disable MD010 MD036 -->
+# NAME
+
+	distrobox ephemeral
+	distrobox-ephemeral
+
+# DESCRIPTION
+
+distrobox-ephemeral creates a temporary distrobox that is automatically destroyed
+when the command is terminated.
+
+# SYNOPSIS
+
+**distrobox ephemeral**
+
+	--root/-r:		launch podman/docker with root privileges. Note that if you need root this is the preferred
+				way over "sudo distrobox"
+	--verbose/-v:		show more verbosity
+	--help/-h:		show this message
+	--/-e:			end arguments execute the rest as command to execute at login	default: bash -l
+	--version/-V:		show version
+
+# EXAMPLES
+
+	distrobox-ephemeral --image alpine:latest -- cat /etc/os-release
+	distrobox-ephemeral --root --verbose --image alpine:latest --volume /opt:/opt
+
+You can also use [flags from **distrobox-create**](distrobox-create.md) to customize the ephemeral container to run.
+
+Refer to
+
+	man distrobox-create
+
+or
+
+	distrobox-create --help
+
+Supported environment variables:
+
+	distrobox-ephemeral calls distrobox-create, SEE ALSO distrobox-create(1) for
+	a list of supported environment variables to use.
diff --git a/docs/usage/usage.md b/docs/usage/usage.md
index 01691ae..14f67be 100644
--- a/docs/usage/usage.md
+++ b/docs/usage/usage.md
@@ -2,6 +2,7 @@
   - [Outside the distrobox](#outside-the-distrobox)
     - [distrobox-create](distrobox-create.md)
     - [distrobox-enter](distrobox-enter.md)
+    - [distrobox-ephemeral](distrobox-ephemeral.md)
     - [distrobox-list](distrobox-list.md)
     - [distrobox-rm](distrobox-rm.md)
     - [distrobox-stop](distrobox-stop.md)
-- 
2.37.0

